Packet++/header/VrrpLayer.h

```diff
@@ -0,0 +1,560 @@
+		* @param[in] data A pointer to the raw data
+		* @param[in] dataLen Size of the data in bytes
+		* @param[in] prevLayer A pointer to the previous layer
+		* @param[in] packet A pointer to the Packet instance where layer will be stored in
+		*/
+		VrrpV2Layer(uint8_t *data, size_t dataLen, Layer *prevLayer, Packet *packet)
+				: VrrpLayer(data, dataLen, prevLayer, packet, VRRPv2, IPAddress::IPv4AddressType) {}
+
+		/**
+		 * A constructor that allocates a new VRRPv2
+		 */
+		VrrpV2Layer() : VrrpLayer(VRRPv2) { setAddressType(IPAddress::IPv4AddressType); };
+
+		/**
+		 * A destructor for this layer (does nothing)
+		 */
+		~VrrpV2Layer() {}
+
+		// implement abstract methods
+
+		/**
+		* @return The VRRP authentication type description
+		*/
+		std::string getAuthTypeDesc() const;
+
+		/**
+		* A method that gets VRRP advertisement interval
+		* @return The advertisement interval in this message
+		*/
+		uint8_t getAdvInt() const;
+
+		/**
+		* A method that gets VRRP authentication type
+		* @return The authentication type in this message
+		*/
+		uint8_t getAuthType() const;
+
+		/**
+		* Fill the checksum from header and data and possibly write the result to @ref vrrp_packet#checksum
+		* @return The checksum result
+		*/
+		uint16_t fillChecksum();
+
+		/**
+		* Calculate the checksum from header and data and possibly write the result to @ref vrrp_packet#checksum
+		* @return The checksum result
+		*/
+		uint16_t calculateChecksum() const;
+	};
+
+	/**
+	 * @class VrrpV3Layer
+	 * Represents VRRPv3 (Virtual Router Redundancy Protocol ver 3) layer. This class represents all the different messages of VRRP
+	 */
+	class VrrpV3Layer : public VrrpLayer {
+	public:
+		/** A constructor that creates the layer from an existing packet raw data
+		* @param[in] data A pointer to the raw data
+		* @param[in] dataLen Size of the data in bytes
+		* @param[in] prevLayer A pointer to the previous layer
+		* @param[in] packet A pointer to the Packet instance where layer will be stored in
+		* @param[in] addressType IP address type
+		*/
+		VrrpV3Layer(uint8_t *data, size_t dataLen, Layer *prevLayer, Packet *packet, IPAddress::AddressType addressType)
+				: VrrpLayer(data, dataLen, prevLayer, packet, VRRPv3, addressType) {}
+
+		/**
+		 * @param[in] addressType IP version of virtual IP address
+		 * A constructor that allocates a new VRRPv3
+		 */
+		explicit VrrpV3Layer(IPAddress::AddressType addressType) : VrrpLayer(
+				VRRPv3) { setAddressType(addressType); };
+
+		/**
+		 * A destructor for this layer (does nothing)
+		 */
+		~VrrpV3Layer() {}
+
+		// implement abstract methods
+
+
+		/**
+		 * A method that gets VRRP authentication type description
+		* @return The VRRP authentication type description
+		*/
+		std::string getAuthTypeDesc() const;
+
+		/**
+		* A method that gets VRRP maximum advertisement interval
+		* @return The maximum advertisement interval in this message
+		*/
+		uint8_t getAdvInt() const;
```

Comment:
If I understand correctly in VRRP v3 the AdvInt size is bigger than `unit8_t` (8 bit), it's actually 12-bit: https://datatracker.ietf.org/doc/html/rfc5798

I think that instead of creating a virtual `getAdvInt()` method in `VrrpLayer` and then override it in `VrrpV2Layer` and `VrrpV3Layer` you can just implement 2 separate methods in both classes, each one with a slightly different signature (in v2 it'll return `uint8_t` and in v3 it'll return `uint16_t`)

Reply:
OK. I will fix it.

Reply:
I'm not sure this method must be on the parent, we can define it separately on `VrrpV2Layer` and `VrrpV3Layer` with a slightly different signature (in v2 it'll return `uint8_t` and in v3 it'll return `uint16_t`)

Reply:
Fixed in https://github.com/seladb/PcapPlusPlus/pull/1118/commits/71468b7f0c5c05bb3f6c6f9cc96c47ebbb448ef7